Index: src/org/apache/tomcat/util/net/jss/JSSSocketFactory.java
===================================================================
--- src/org/apache/tomcat/util/net/jss/JSSSocketFactory.java	(revision 278)
+++ src/org/apache/tomcat/util/net/jss/JSSSocketFactory.java	(working copy)
@@ -405,6 +405,10 @@
             // MUST look for "clientauth" (ALL lowercase) since "clientAuth"
             // (camel case) has already been processed by Tomcat 7
             String clientAuthStr = (String)endpoint.getAttribute("clientauth");
+            if (clientAuthStr == null) {
+                debugWrite("JSSSocketFactory init - \"clientauth\" not found, default to want.");
+                clientAuthStr = "want";
+            }
             File file = null;
             try {
                 mServerCertNickPath = (String)endpoint.getAttribute("serverCertNickFile");
